src/write.rs

```diff
@@ -124,65 +124,198 @@ fn verify_channel_mask() {
     assert_eq!(channel_mask(4), 15);
 }
 
-/// A writer that accepts samples and writes the WAVE format.
-///
-/// The writer needs a `WavSpec` that describes the audio properties. Then
-/// samples can be written with `write_sample`. Channel data is interleaved.
-/// The number of samples written must be a multiple of the number of channels.
-/// After all samples have been written, the file must be finalized. This can
-/// be done by calling `finalize`. If `finalize` is not called, the file will
-/// be finalized upon drop. However, finalization may fail, and without calling
-/// `finalize`, such a failure cannot be observed.
-pub struct WavWriter<W>
-    where W: io::Write + io::Seek
-{
-    /// Specifies properties of the audio data.
-    spec: WavSpec,
+enum FmtKind {
+    PcmWaveFormat,
+    WaveFormatExtensible,
+}
 
-    /// The (container) bytes per sample. This is the bit rate / 8 rounded up.
-    bytes_per_sample: u16,
+pub struct EmbeddedWriter<'w, W: 'w + io::Write + io::Seek> {
+    writer: &'w mut W,
+    state: ChunkWritingState,
+}
 
-    /// The writer that will be written to.
+impl<'w, W: 'w + io::Write + io::Seek> Write for EmbeddedWriter<'w, W> {
+    fn write(&mut self, buf:&[u8]) -> io::Result<usize> {
+        let &mut EmbeddedWriter { ref mut writer, ref mut state } = self;
+        state.write(writer, buf)
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        let &mut EmbeddedWriter { ref mut writer, ref mut state } = self;
+        try!(state.update_header(writer));
+        writer.flush()
+    }
+}
+
+impl<'w, W: 'w + io::Write + io::Seek> Drop for EmbeddedWriter<'w, W> {
+    fn drop(&mut self) {
+        let &mut EmbeddedWriter { ref mut writer, ref mut state } = self;
+        let _ignored = state.finalize_chunk(writer);
+    }
+}
+
+#[derive(Copy,Clone)]
+pub struct ChunkWritingState {
+    pub len: u32
+}
+
+impl ChunkWritingState {
+    pub fn write<W: io::Write + io::Seek>(&mut self, writer: &mut W, buf:&[u8]) -> io::Result<usize> {
+        let written = try!(writer.write(buf));
+        self.len += written as u32;
+        Ok(written)
+    }
+
+    pub fn update_header<W: io::Write + io::Seek>(&mut self, writer: &mut W) -> io::Result<()> {
+        try!(writer.seek(io::SeekFrom::End(-(self.len as i64 + 4))));
```

Comment:
Shouldn’t this be `SeekFrom::Current`? `End` might go wrong when the underlying writer started at some offset before the end and when writing the chunk did not reach the end yet.